### PR TITLE
Realtime: Add `domain` and `secure` to config

### PR DESCRIPTION
**Pull request by gregorylegarec**
~~BREAKING CHANGE: `url` is no more used in config and must be replaced by `domain` and `secure` attributes~~

~~Two informations was passed with the `url` attribute: the security protocol and the cozy domain. But they're was a flaw with url like `http://cozy.tools:8080`. The unsecure protocol was detected correctly, but the domain was not parsed and has `undefined` value.~~

Eventually, this PR just adresses the issue caused by a bad RegExp matching.

### CODE REVIEW

**Comment by gregorylegarec** [comment]
[[STO] Initialisation du realtime échoue](https://trello.com/c/j5oi8c1J/1171-sto-initialisation-du-realtime-%C3%A9choue)

**Comment by kosssi** [comment]
@gregorylegarec By default I prefer that it's in secure mode

File: packages/realtime/package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "cozy-realtime",
-  "version": "1.1.2",
+  "version": "2.0.0",
```
Comment (kosssi) [inline packages/realtime/package.json]: You do not have to change the version numbers ;)

**Comment by ptbrowne** [comment]
> The unsecure protocol was detected correctly, but the domain was not parsed and has undefined value.

Why the API change to resolve for this bug ? Why not parse the domain correctly ? We are accustomed to use the URL property to configure libraries communicating with Cozy.

**Comment by gregorylegarec** [comment]
@ptbrowne The thing is we discussed with @kosssi and agreed to update the way realtime was initialized, following my proposition.

Why changing ? For at least two reasons:
* The stack send us `domain` and `token`, I think we should keep them unchanged, having to add a `http(s)` protocol to the domain is pretty much useless, especially for the realtime where this protocol is just tested with a regexp to use `ws(s)` instead (Actually, that's the current regexp who fails to match the domain an causes the current bug.
* In my opinion, it is always a code smell to pass more than one information in a string, and it's exactly what we are doing here: by adding a protocol to the `domain` given by the stack to make the `url` variable, we are juste telling our library : "Hey, look at this protocol to determine if you have to use SSL and yes also this the domain". I would prefer to make everything explicit, so that's why I would prefer to add a `secure` property.

After your comment I discussed with @kosssi and we agreed it is eventually too soon to make all those changes, so those two points should be adressed in a dedicated issue about how are we initializing all ou libraries. I will create it in a few minutes.

So for now I will update this PR to just fix the behaviour which make the realtime unusable in a local environment.

**Comment by ptbrowne** [comment]
Thanks Greg for your detailed explanation. I understand better now. Cool that the breaking change will be done in another issue and that first the bug will be fixed.

Concerning the API : I think we should not have to pass `secure`. This makes the API larger on the application side and it is a burden on the user to decide whether to use `secure` or not. The underlying library can choose whether or not to do SSL based on the `window.location.protocol`.

Concerning the first bug : Parsing URLs with a RegExp is a bad idea, it has already bitten us once for the intents when the stack started to put hash fragments in its URLs. The URL API seems to be a good choice here.

image

**Comment by gregorylegarec** [comment]
@ptbrowne Totally agree with an internal use of `window.location.protocal`, that was how the first version of the realtime lib was proceeding back in the days. BUT I have a mix of both `window.location.protocol` and `secure` attribute in my mind, but I will detail it later to discuss about it.

I thank about the `URL` API but I think it is not available in Node for desktop and mobile apps using realtime.

**Comment by ptbrowne** [comment]
The URL API is available in NodeJS 8 : https://nodejs.org/docs/latest-v8.x/api/url.html

**Comment by gregorylegarec** [comment]
Nice, so let's go.

**Comment by gregorylegarec** [comment]
@ptbrowne Ok code has been updated. I added a type check for `url` and `token` parameter because I have been stuck an hour because of an Object token (`cozy.client._token` being an object).

File: packages/realtime/src/index.js [hunk 2]
```diff
@@ -75,6 +77,15 @@ async function connectWebSocket(
...
   retryDelay
 ) {
+  const expectedStringAttributes = ['domain', 'token']
+  expectedStringAttributes.forEach(attribute => {
+    if (typeof config[attribute] !== 'string') {
+      throw new Error(
+        `Configuration attribute ${attribute} should be a string.`
+      )
+    }
+  })
```
Comment (ptbrowne) [inline packages/realtime/src/index.js]: Minor: for this kind of code I usually go in a PropTypes manner.

```js
const isString = [
  str => typeof str === 'string',
  'should be a string'
]
const configTypes = {
  domain: isString,
  token: isString
}
const validate = config => obj => {
	for (const [attr, [validator, message]] of Object.items(configTypes)) {
      if (!validator(obj[attr])) {
        throw new Error('Attribute ${attribute} ${message}')
      }
    }
    return true
}
const validateConfig = validate(configTypes)
```

Comment (gregorylegarec) [reply packages/realtime/src/index.js]: I used this pattern to validate, thanks. I let you merge if it is ok.

**Pull request merged**